smartcontractkit/external-adapters-js
# [Perf] Deterministic Cache keys

## Description
To avoid `object-hash` sorts we can instead use the `inputParameters` of a request to generate a deterministic cache key by hashing together the values.

## Changes

- Introduces a new middleware `cache-key` that is used to generate the cache key one time, then add it to the `AdapterContext` for re-use across the EA framework's execution.
- Modifies the structure of batch data `.results` to include the child data's cache key. Updates batchable EAs with this change.

## Caveats
- WS cache keys still use `object-hash`ing

## TODO:
- [ ] WS cache keys
- [ ] Additional unit tests
- [ ] Profiling



## Steps to Test
Single pair
1. Run any external adapter with batch warming
2. Send un-batched single pair request
3. Request again. It should be cached from the request on step 2.

Single pair from warmer
1. Run any external adapter with batch warming
2. Send un-batched single pair request
3. Request again after ~1 minute. It should be cached after being topped up from the cache warmer.

Batch request
1. Run any external adapter with batch warming
2. Send batched request (e.g. `base: ["BTC", "ETH"], quote: "USD"`)
3. Request again. It should be cached from the request on step 2.

Batch request from warmer
1. Run any external adapter with batch warming
2. Send batched request (e.g. `base: ["BTC", "ETH"], quote: "USD"`)
3. Request again after ~1 minute. It should be cached after being topped up from the cache warmer.

Single pairs fulfilled from warmer
1. Run any external adapter with batch warming
2. Send multiple single pair requests
3. Request each of them again after ~1 minute. It should be cached after being topped up from the cache warmer.

## Quality Assurance
